dwarf-rust-experimental/rust/src/main.rs

```diff
@@ -0,0 +1,502 @@
+                  Err(_) => None,
+                }
+              }
+              gimli::AttributeValue::String(name_str) => {
+                Some(name_str.to_string()?.to_string())
+              }
+              _ => None,
+            }
+          } else {
+            None
+          };
+          
+
+          // Get low_pc (start address) and high_pc (end address or size)
+          let low_pc = if let Some(attr) = entry.attr_value(gimli::DW_AT_low_pc)? {
+            match attr {
+              gimli::AttributeValue::Addr(addr) => Some(addr),
+              _ => None,
+            }
+          } else {
+            None
+          };
+
+          let high_pc = if let Some(attr) = entry.attr_value(gimli::DW_AT_high_pc)? {
+            match attr {
+              gimli::AttributeValue::Addr(addr) => Some(addr),
+              gimli::AttributeValue::Udata(offset) => {
+                // high_pc can be an offset from low_pc
+                if let Some(low) = low_pc {
+                  Some(low + offset)
+                } else {
+                  None
+                }
+              }
+              _ => None,
+            }
+          } else {
+            None
+          };
+
+          // If we have all the information, add it to our functions map
+          if let (Some(func_name), Some(start), Some(end)) = (func_name, low_pc, high_pc) {
+            // Filter out synthetic/relocated addresses similar to line processing
+            const MAX_REALISTIC_WASM_ADDR: u64 = 0x40000000; // 1GB threshold
+            if start <= MAX_REALISTIC_WASM_ADDR && end <= MAX_REALISTIC_WASM_ADDR && start < end {
+              functions.insert(func_name, vec![start, end]);
+            }
+          }
+        }
+      }
+    }
+
+    let mut scored_source_files: IndexMap<(Option<String>, String), ScoredSourceFile> =
+      IndexMap::new();
+
+    let mut rows = line_program.rows();
+
+    // let mut func_state = FuncState::Start;
+
+    while let Some((header, row)) = rows.next_row()? {
+      // if let FuncState::Start = func_state {
+      //   func_state = if row.address() == 0 {
+      //     FuncState::Ignored
+      //   } else {
+      //     FuncState::Normal
+      //   };
+      // }
+      // if let FuncState::Ignored = func_state {
+      //   if row.end_sequence() {
+      //     func_state = FuncState::Start;
+      //   }
+      //   continue;
+      // }
+      // if row.end_sequence() {
+      //   func_state = FuncState::Start;
+      // }
+
+      let file = match row.file(header) {
+        Some(file) => file,
+        None => continue,
+      };
+
+      let pos = {
+        let mut line = match row.line() {
+          Some(line) => line.checked_sub(1).unwrap().try_into().unwrap(),
+          None => continue, // couldn't attribute instruction to any line
+        };
+
+        // It seems we need to add 1 to the line numbers for Rust.
+        let is_rust = lang == constants::DW_LANG_Rust.0;
```

Comment:
is this intended to be Rust-specific? are we going to provider parsers for other languages? how does it work for Go, C etc...